collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -1218,17 +1217,12 @@ private View createView(int event, boolean advancingPage) {
                 final EditText saveAs = endView.findViewById(R.id.save_name);
 
                 // disallow carriage returns in the name
-                InputFilter returnFilter = new InputFilter() {
-                    public CharSequence filter(CharSequence source, int start,
-                                               int end, Spanned dest, int dstart, int dend) {
-                        String newText = source.toString().substring(start, end);
-                        String invalidCharRegex = "[\\p{Cntrl}]";
-
-                        // Replace invalid characters, only modifying the string if necessary.
-                        return newText.matches(invalidCharRegex)
-                                ? newText.replaceAll(invalidCharRegex, " ")
-                                : null;
-                    }
+                InputFilter returnFilter = (source, start, end, dest, dstart, dend) -> {
+                    String newText = source.toString().substring(start, end);
+                    String invalidCharRegex = "[\\p{Cntrl}]";
+
+                    // Replace invalid characters, only modifying the string if necessary.
+                    return newText.replaceAll(invalidCharRegex, " ");
```

Comment:
I don't understand why this logic needed to change, returning `null` here signifies that the string should be unmodified. This change seems like it would hurt performance without actually changing the behavior. Can you give more context for why this works please?

Reply:
Your code almost always returned null because you used `newText.matches(invalidCharRegex)`
If you paste a text eg. `123\nfghgfh` it doesn't match `[\\p{Cntrl}]` of course, the same if the form name is created based on answers like in the sample form.

Reply:
So newlines were never returned it looked like it's in the same line because we use single line EditText.

Reply:
Oh weird, [apparently](https://stackoverflow.com/a/8923446/763231) Java has a non-standard implementation of regex matchers. It should be `invalidCharRegex = ".*[\\p{Cntrl}].*"`. What do you think about updating the regex and keeping the `return null` performance improvement around?

Reply:
It won't work. You will be able to copy\pase a string with newlines as a form name.

Reply:
All I'm suggesting is that `returnFilter` should return `null` if `invalidCharRegex` doesn't match `newText`. If it doesn't match, then of course `newText.replaceAll(invalidCharRegex)` will have no effect because there will be no replacements. It shouldn't change the behavior right?

Reply:
I understand what you mean but the REGEX you meantioned is wrong:
If I have a string with one new line char eg: `aaa\naaaa` mathces() returns true but if I have two new line chars eg `aaa\n\naaa` it returns false. I could improve it, this one seems to work well `.*[\\p{Cntrl}]{1,}..*` but do you really think it's faster? I think my approach - trying to replace new line chars everytime is safer and clearer in this case.

Reply:
Damn 🤦‍♂️ Java is a nightmare sometimes. I tried again (here's [my working example code](https://repl.it/repls/GrimyUncomfortablePlane) using an online REPL) and got a much better solution that actually works.

It does make a big impact on performance; string **matching** is quick but string **assignment** is not, so returning an unchanged string from this method causes the UI to block on unnecessary string assignments. Not a big deal most of the time, but type quickly and you'll get lag after every character you type for large strings.

My updated proposal is basically this:

```java
    Pattern p = Pattern.compile("[\\p{Cntrl}]", Pattern.MULTILINE);
    Matcher m = p.matcher(newText);
    return m.find() ? m.replaceAll(" ") : null;
```

Reply:
Ok done, but I don't think we need that second parameter `Pattern.MULTILINE` am I wrong?